src/Features/Core/Portable/CustomMessageHandler/CustomMessageHandlerService.cs

```diff
@@ -0,0 +1,182 @@
+
+        Document? document = null;
+        if (documentId is not null)
+        {
+            document = solution.GetDocument(documentId) ?? await solution.GetSourceGeneratedDocumentAsync(documentId, cancellationToken).ConfigureAwait(false);
+            Contract.ThrowIfNull(document);
+        }
+
+        try
+        {
+            ICustomMessageHandlerWrapper? handler;
+            lock (_lockObject)
+            {
+                // Check if the assembly is already loaded.
+                if (!_extensions.TryGetValue(assemblyFolderPath, out var extension))
+                {
+                    var loadContext = new System.Runtime.Loader.AssemblyLoadContext(name: $"RemoteCustomMessageHandlerService assembly load context for {assemblyFolderPath}", isCollectible: true);
+                    loadContext.Resolving += ResolveExtensionAssembly;
+
+                    extension = new CustomMessageHandlerExtension(loadContext);
+                    _extensions[assemblyFolderPath] = extension;
+                }
+
+                if (!extension.Handlers.TryGetValue(typeFullName, out handler))
+                {
+                    var assembly = extension.AssemblyLoadContext.LoadFromAssemblyPath(assemblyPath);
+                    var type = assembly.GetType(typeFullName)
+                        ?? throw new InvalidOperationException($"Cannot find type {typeFullName} in {assemblyPath}.");
+
+                    handler = _customMessageHandlerFactory.Create(type);
+                }
+            }
+
+            var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+
+            var result = await handler.ExecuteAsync(message, document, solution, cancellationToken)
+                .ConfigureAwait(false);
+
+            var responseJson = JsonSerializer.Serialize(result, handler.ResponseType);
+
+            return responseJson;
+        }
+        catch (Exception e)
+        {
+            Logger.Log(
+                FunctionId.CustomMessageHandlerService_HandleCustomMessageAsync,
+                $"Custom handler {assemblyPath} {typeFullName} error: {e}",
+                LogLevel.Error);
+            throw;
+        }
+
+        Assembly? ResolveExtensionAssembly(System.Runtime.Loader.AssemblyLoadContext context, AssemblyName assemblyName)
+        {
+            try
+            {
+                return _defaultLoadContext.LoadFromAssemblyName(assemblyName);
+            }
+            catch
+            {
+            }
+
+            var extensionAssemblyPath = Path.Combine(assemblyFolderPath, $"{assemblyName.Name}.dll");
+
+            // This will throw FileNotFoundException if the assembly is not found.
+            return context.LoadFromAssemblyPath(extensionAssemblyPath);
+        }
+    }
+
+    public ValueTask UnloadCustomMessageHandlersAsync(
+        string assemblyFolderPath,
+        CancellationToken cancellationToken)
+    {
+#if DEBUG
```

Comment:
Remove

Reply:
I guess this comment refers to
```
System.Diagnostics.Debugger.Launch();
```
These are very convenient for my debugging right now. I will definitely remove them all before completing the PR. I will keep this comment open until then.